python/selfie-lib/selfie_lib/ArrayMap.py

```diff
@@ -1,27 +1,41 @@
 from collections.abc import Set, Iterator, Mapping
-from typing import List, TypeVar, Union, Any
+from typing import List, TypeVar, Union, Any, Callable, Optional, Generator
 from abc import abstractmethod, ABC
-from functools import total_ordering
 
 T = TypeVar("T")
 V = TypeVar("V")
 K = TypeVar("K")
 
 
-@total_ordering
-class Comparable:
-    def __init__(self, value):
-        self.value = value
+class BinarySearchUtil:
+    @staticmethod
+    def binary_search(
+        data, item, compare_func: Optional[Callable[[Any, Any], int]] = None
+    ) -> int:
```

Comment:
1. since we're in Python, this doesn't *have* to be in a class. It could just be `def _binary_search_util(data, item):`

2. This function is only ever called with two parameters -`compare_func` is always None. The fix is

```python
def _compare_normal(a, b) -> int:
  if a == b:
    return 0
  else if a < b:
    return -1
  else
    return 1

def _compare_string_slash_first(a: str, b : str) -> int:
  return _compare_normal(a.replace("/", "\0"), b.replace("/", "\0"))

def _binary_search(data, item) -> int:
  compare_func = _compare_string_slash_first if item is str else _compare_normal
  low, high = 0, len(data) - 1
  ...
  the rest of the logic, but now you know that compare_func is never None
```